lib/browser/auth.ts

```diff
@@ -11,17 +11,305 @@
+                        this.expire_time = this.source_provider.expireTime;
+                    }
+                });
+            }
+            else
+            {
+                this.aws_credentials.aws_access_id = this.source_provider.accessKeyId;
+                this.aws_credentials.aws_secret_key = this.source_provider.secretAccessKey;
+                this.aws_credentials.aws_sts_token = this.source_provider.sessionToken;
+                this.aws_credentials.aws_region = (this.source_provider_options as AWSCognitoCredentialOptions).Region;
+                this.expire_time = this.source_provider.expireTime;
+
+            }
+       });
+    }
+
+
+    async refreshCredentialAsync()
+    {
+        return new Promise<AWSCognitoCredentialsProvider>((resolve, reject) => {
+            this.source_provider.get((err)=>{
+                if(err)
+                {
+                    reject("Failed to get cognito credentials.")
+                }
+                else
+                {
+                    this.aws_credentials.aws_access_id = this.source_provider.accessKeyId;
+                    this.aws_credentials.aws_secret_key = this.source_provider.secretAccessKey;
+                    this.aws_credentials.aws_sts_token = this.source_provider.sessionToken;
+                    this.aws_credentials.aws_region = (this.source_provider_options as AWSCognitoCredentialOptions).Region;
+                    this.expire_time = this.source_provider.expireTime;
+                    resolve(this);
+                }
+            });
+        });
+    }
+}
+
+
+
+/**
+ * AWSCredentialsProviderCached. The AWSCredentialsProviderCached will be our main provider class.
+ * The AWSCredentialsProviderCached will cached the current credential and expired_time, and stores a list of credentials providers. 
+ *          If the credential is not expired, return the cached credentials.
+ *          If the credential is expired, refresh the credential, and check the next credentialsProvider.
+ * @category Auth
+ */
+export class AWSCredentialsProviderCached extends CredentialsProvider{
+    cached_credentials : AWSCredentials | undefined;
+    source_provider : CredentialsProvider | null;
+
+    constructor(options: CredentialsOptions, expire_interval_in_ms? : number)
+    {
+        super(options,expire_interval_in_ms);
+        var provider = null;
+        if(options instanceof StaticCredentialOptions)
+        {
+            provider = new StaticCredentialProvider(options, expire_interval_in_ms);
+        }
+        else if (options instanceof AWSCognitoCredentialOptions)
+        {
+            provider = new AWSCognitoCredentialsProvider(options, expire_interval_in_ms);
+        }
+        this.source_provider = provider;
+        this.cached_credentials = this.source_provider?.getCredentials();
+        this.expire_time = this.source_provider?.expire_time;
+    }
+
+    add_provider(provider: CredentialsProvider)
```

Comment:
not sure what's going on here

Reply:
This function allow us to add multiple providers. The cached provider will keep a chain of the credentialsprovider and retrieved credentials from the provider chain, if the cached credentials expired.

Reply:
Caches and chain are two separate concerns; we should stick to one.